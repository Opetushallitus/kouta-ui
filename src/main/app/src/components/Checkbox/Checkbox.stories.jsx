import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Checkbox from './index';

storiesOf('Checkbox', module)
  .add('Basic', () => (
    <Checkbox checked={true} onChange={action('change')}>
      Checkbox
    </Checkbox>
  ))
  .add('With disabled', () => (
    <Checkbox checked={true} onChange={action('change')} disabled>
      Checkbox
    </Checkbox>
  ));
