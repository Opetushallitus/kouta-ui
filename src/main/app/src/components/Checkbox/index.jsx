import React from 'react';
import styled, { css } from 'styled-components';

import { getThemeProp } from '../../theme';

const CheckboxContainer = styled.div`
  display: inline-block;
  position: relative;
  top: 0.15em;
`;

const Icon = styled.svg`
  fill: none;
  stroke: white;
  stroke-width: 2px;
`;
const HiddenCheckbox = styled.input.attrs({ type: 'checkbox' })`
  border: 0;
  clip: rect(0 0 0 0);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

const StyledCheckbox = styled.div`
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid ${getThemeProp('palette.border')};
  box-shadow: 0 0 2px 0 rgba(0, 0, 0, 0), inset 0 1px 2px 0 rgba(0, 0, 0, 0.1);
  border-radius: 2px;
  transition: box-shadow 0.25s, border-color 0.25s, background-color 0.25s;

  ${HiddenCheckbox}:focus + & {
    border-color: ${getThemeProp('palette.primary.main')};
    box-shadow: 0 0 0 1px ${getThemeProp('palette.primary.main')};
  }

  ${({ checked }) =>
    checked &&
    css`
      background-color: ${getThemeProp('palette.primary.main')};
      border-color: ${getThemeProp('palette.primary.main')};
    `}

  ${({ error }) =>
    error &&
    css`
      border-color: ${getThemeProp('palette.danger.main')};
    `}

  ${Icon} {
    visibility: ${props => (props.checked ? 'visible' : 'hidden')};
  }
`;

const Label = styled.label`
  font-family: ${getThemeProp('typography.fontFamily')};
  font-size: 1rem;
  display: flex;
  line-height: 1.5;
  color: ${getThemeProp('palette.text.primary')};
  align-items: flex-start;

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
      cursor: not-allowed;
    `}
`;

const LabelWrapper = styled.div`
  flex: 1;
  margin-left: 9px;

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`;

const CheckboxWrapper = styled.div`
  flex: 0;
  line-height: 0;
`;

const IndeterminateIndicator = styled.div`
  top: 50%;
  left: 50%;
  width: 0.5em;
  height: 0.5em;
  background-color: ${({ theme }) => theme.palette.primary.main};
  border-radius: 2px;
  position: absolute;
  transform: translate(-50%, -50%);
`;

const Checkbox = ({
  className,
  checked,
  children,
  error = false,
  disabled = false,
  indeterminate = false,
  ...props
}) => (
  <Label disabled={disabled}>
    <CheckboxWrapper>
      <CheckboxContainer className={className}>
        <HiddenCheckbox checked={checked} disabled={disabled} {...props} />
        <StyledCheckbox
          checked={checked}
          indeterminate={indeterminate}
          error={error}
        >
          <Icon viewBox="0 0 24 24">
            <polyline points="20 6 9 17 4 12" />
          </Icon>
          {indeterminate && !checked ? <IndeterminateIndicator /> : null}
        </StyledCheckbox>
      </CheckboxContainer>
    </CheckboxWrapper>
    <LabelWrapper>{children}</LabelWrapper>
  </Label>
);

export default Checkbox;
