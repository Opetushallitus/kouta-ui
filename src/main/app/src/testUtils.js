import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const shallowWithTheme = makeFnWithTheme(shallow);

export const mountWithTheme = makeFnWithTheme(mount);

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => +date;

  return () => {
    global.Date = _Date;
  };
};
