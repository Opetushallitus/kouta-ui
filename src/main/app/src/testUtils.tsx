import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';

import { createRootReducer } from './state/rootReducer';
import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const renderWithTheme = makeFnWithTheme(render);

export const mountWithTheme = tree => {
  const { container } = makeFnWithTheme(render)(tree);
  return container.firstChild;
};

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => Number(date);

  return () => {
    global.Date = _Date;
  };
};

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: createRootReducer({}), preloadedState }),
    ...renderOptions
  }
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
      </Provider>
    );
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
