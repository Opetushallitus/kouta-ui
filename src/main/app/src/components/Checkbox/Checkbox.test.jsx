import React from 'react';

import Checkbox from './index';
import { mountWithTheme } from '../../testUtils';

test('renders correctly', () => {
  expect(mountWithTheme(<Checkbox>Checkbox</Checkbox>)).toMatchSnapshot();
});
