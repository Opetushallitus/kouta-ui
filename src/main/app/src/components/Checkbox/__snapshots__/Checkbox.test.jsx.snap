// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2 {
  display: inline-block;
  position: relative;
}

.c7 {
  fill: none;
  stroke: white;
  stroke-width: 2px;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c5 {
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #cccccc;
  box-shadow: 0 0 2px 0 rgba(0,0,0,0),inset 0 1px 2px 0 rgba(0,0,0,0.1);
  border-radius: 2px;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
}

.c3:focus + .c5 {
  border-color: #2da0c7;
  box-shadow: 0 0 0 1px #2da0c7;
}

.c5 .c6 {
  visibility: hidden;
}

.c0 {
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.5;
  color: #666666;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

.c1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
}

<styled.label
  disabled={false}
>
  <label
    className="c0"
    disabled={false}
  >
    <styled.div>
      <div
        className="c1"
      >
        <styled.div>
          <div
            className="c2"
          >
            <styled.input
              disabled={false}
            >
              <input
                className="c3 c4"
                disabled={false}
                type="checkbox"
              />
            </styled.input>
            <styled.div
              error={false}
              indeterminate={false}
            >
              <div
                className="c5"
              >
                <styled.svg
                  viewBox="0 0 24 24"
                >
                  <svg
                    className="c6 c7"
                    viewBox="0 0 24 24"
                  >
                    <polyline
                      points="20 6 9 17 4 12"
                    />
                  </svg>
                </styled.svg>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </div>
    </styled.div>
    <styled.div>
      <div
        className="c8"
      >
        Checkbox
      </div>
    </styled.div>
  </label>
</styled.label>
`;
